lib/services/android-project-service.ts

```diff
@@ -118,19 +119,135 @@ class AndroidProjectService implements IPlatformProjectService {
+            if (match) {
+                var libRef: ILibRef = { idx: parseInt(match[1]), path: match[2].trim() };
+                libRef.adjustedPath = thiz.$fs.isRelativePath(libRef.path) ? path.join(projDir, libRef.path) : libRef.path;
+                thiz.parseProjectProrperies(libRef.adjustedPath, destDir);
+            }
+        });
+
+        this.$logger.info("Copying %s", projDir);
+        shell.cp("-Rf", projDir, destDir);
+
+        var targetDir = path.join(destDir, path.basename(projDir));
+        // TODO: parametrize targetSdk
+        var targetSdk = "android-17";
+        this.generateBuildFile(targetDir, targetSdk);
+    }
+
+    private getProjectReferences(projDir: string): ILibRef[]{
+        var projProp = path.join(projDir, "project.properties");
+
+        var lines = fs.readFileSync(projProp, { encoding: "utf-8" }).split("\n");
+        var thiz = this;
+
+        var refs: ILibRef[] = [];
+
+        lines.forEach((elem, idx, arr) => {
+            var match = elem.match(/android\.library\.reference\.(\d+)=(.*)/);
+            if (match) {
+                var libRef: ILibRef = { idx: parseInt(match[1]), path: match[2] };
+                libRef.adjustedPath = path.join(projDir, libRef.path);
+                refs.push(libRef);
+            }
+        });
+
+        return refs;
+    }
+
+    private updateProjectReferences(projDir: string, libraryPath: string): void {
+        var refs = this.getProjectReferences(projDir);
+        var maxIdx = refs.length > 0 ? _.max(refs, r => r.idx).idx : 0
+
+        var relLibDir = path.relative(projDir, libraryPath).split("\\").join("/");
+
+        var libRefExists = _.filter(refs, r => path.normalize(r.path) == path.normalize(relLibDir)).length > 0;
+
+        if (!libRefExists) {
+            var projRef = util.format("\nandroid.library.reference.%d=%s", maxIdx + 1, relLibDir);
+            var projProp = path.join(projDir, "project.properties");
+            fs.appendFileSync(projProp, projRef, { encoding: "utf-8" });
+        }
+    }
+
+    public addLibrary(platformData: IPlatformData, libraryPath: string): IFuture<void> {
+        var name = path.basename(libraryPath);
+        var projDir = this.$projectData.projectDir;
+        var targetPath = path.join(projDir, "lib", platformData.normalizedPlatformName);
+        this.$fs.ensureDirectoryExists(targetPath).wait();
+
+        this.parseProjectProrperies(libraryPath, targetPath);
+
+        shell.cp("-f", path.join(libraryPath, "*.jar"), targetPath);
+        var projectLibsDir = path.join(platformData.projectRoot, "libs");
+        this.$fs.ensureDirectoryExists(projectLibsDir).wait();
+        shell.cp("-f", path.join(libraryPath, "*.jar"), projectLibsDir);
+
+        var targetLibPath = path.join(targetPath, path.basename(libraryPath));
+
+        this.updateProjectReferences(platformData.projectRoot, targetLibPath);
+        return Future.fromResult();
```

Comment:
instead of returning ```Future.fromResult()``` it's better to wrap whole body of the method inside 
```Typescript
return (() => { /*body here*/ }).future<void>()();
```